File: src/dvd/dvd.c
#ifndef DVD_C_
#define DVD_C_

#include "raylib.h"
#include "raymath.h"
#include "../ui/raygui_incl.h"
#include "../gamestate.h"
#include "../const.h"
#include "../ui/font.c"

//! All of the code for rendering and handling the DvD 'game'.

INCBIN(dvd_logo, "assets/DVD_logo.png");

typedef struct Dvd {
    GameState state;
    Vector2 position;
    Vector2 velocity;
    Texture2D logo;
    Vector2 logo_size;
    bool inverted, rainbow;
    float passed_time;
    int key;
} Dvd;

Dvd* dvd_alloc() {
    Dvd* d = calloc(1, sizeof(Dvd));
    d->state = GameState_Help;

    Image logo_png = LoadImageFromMemory(".png", dvd_logo_data, dvd_logo_size);
    d->logo = LoadTextureFromImage(logo_png);
    UnloadImage(logo_png);

    d->logo_size = vec2(d->logo.width, d->logo.height);
    d->velocity  = vec2(250, 250);
    d->inverted  = true;
    d->rainbow   = true;
    d->position  = vec2(
        GetRandomValue(0, GetScreenWidth()  - d->logo.width),
        GetRandomValue(0, GetScreenHeight() - d->logo.height)
    );
    return d;
}

void dvd_free(Dvd* d) {
    UnloadTexture(d->logo);
    free(d);
}

static inline void dvd_set_random_pos(Dvd* d) {
    d->position = vec2(
        GetRandomValue(0, global_state.screen_w - d->logo.width),
        GetRandomValue(0, global_state.screen_h - d->logo.height)
    );
}

static inline void dvd_draw_help() {
    static float spacing;
    spacing = FONT_M;
        
    BeginDrawing();
    ClearBackground(BLACK);

    DrawTextD("DvD controls", 3, FONT_M, FONT_XL, GREEN);
    DrawTextD("H         - toggle help mode", 3, spacing += FONT_XL, FONT_M, RAYWHITE);
    DrawTextD("I          - invert the colors", 3, spacing += FONT_M, FONT_M, RAYWHITE);
    DrawTextD("R         - randomize the DvD logo position", 3, spacing += FONT_M, FONT_M, RAYWHITE);
    DrawTextD("C         - toggle rainbow colors", 3, spacing += FONT_M, FONT_M, RAYWHITE);
    DrawTextD("Space   - pause/unpause the DvD logo", 3, spacing += FONT_M, FONT_M, RAYWHITE);
}

SelectedGame dvd_update(Dvd* d) {
    static float dt;

    static int key;
    key = GetKeyPressed();

    switch (key) {
        case KEY_ESCAPE: {
            d->state = GameState_Paused;
            return Selected_None;   
        }
        case KEY_H: {
            d->state = (d->state == GameState_Help)? GameState_Paused : GameState_Help;
        } break;
        case KEY_I: {
            d->inverted = !(d->inverted);
        } break;
        case KEY_R: {
            dvd_set_random_pos(d);
        } break;
        case KEY_C: {
            d->rainbow = !(d->rainbow);
        } break;
        case KEY_SPACE: {
            d->state = (d->state == GameState_Running)? GameState_Paused : GameState_Running;
        } break;
        case KEY_F5: {
            global_state.show_fps = !global_state.show_fps;
        } break;
        default: {}
    }
    
    dt = GetFrameTime();
    d->passed_time = (d->passed_time > 36.0f)? 0.0f : d->passed_time + dt;

    switch (d->state) {
        case GameState_Help: 
            dvd_draw_help();
            return Selected_DVD;
        case GameState_Running: goto update_dvd;
        case GameState_Paused: goto draw_dvd;
        default: {}
    }

update_dvd:
    d->position = Vector2Add(
        Vector2Multiply(
            (Vector2){dt, dt},
            d->velocity
        ),
        Vector2Clamp(
            d->position,
            (Vector2){0, 0},
            Vector2Subtract(
                vec2(global_state.screen_w, global_state.screen_h),
                d->logo_size
            )
        )
    );

    if (
        d->position.x + d->logo_size.x >= global_state.screen_w ||
        d->position.x <= 0
    ) d->velocity.x *= -1;
    
    if (d->position.y + d->logo_size.y >= global_state.screen_h ||
        d->position.y <= 0
    ) d->velocity.y *= -1;

    draw_dvd: {
        Color tint = d->rainbow?
            d->inverted?
                ColorFromHSV(d->passed_time * 10.0f, 1.0, 1.0) :
                ColorFromHSV(d->passed_time * 10.0, 0.6, 1.0)
            : d->inverted? WHITE : BLACK;

        BeginDrawing();
        ClearBackground((d->inverted)? BLACK : RAYWHITE);
        if (d->state == GameState_Paused) {
            DrawTextD("[SPACE TO UNPAUSE]", 3, global_state.screen_h - FONT_S, FONT_S, GREEN);
        }
        DrawTexture(
            d->logo,
            d->position.x,
            d->position.y,
            tint
        );
    }

    return Selected_DVD;
}

#endif

File: Legacy-rust-version/src/dvd/dvd.rs
use rust_raylib::{
    math::Vector2 as Vec2,
    texture::Texture2D,
    ffi
};

use crate::selector::SelectedGame;

const DVD_LOGO_BYTES: &[u8] = include_bytes!("../assets/DVD_logo.png");

pub struct Dvd {
    pos: Vec2,
    velocity: Vec2,
    window_size: Vec2,
    logo: Texture2D,
    logo_size: Vec2,
}

impl crate::game::Game for Dvd {
    fn new() -> Self {
        let window_size = Vec2{x};

        let logo_raw = unsafe {
            let img = ffi::LoadImageFromMemory(
                ".png".as_ptr() as *const i8, DVD_LOGO_BYTES.as_ptr(),
                DVD_LOGO_BYTES.len() as i32
            );
            let l = ffi::LoadTextureFromImage(img);
            ffi::UnloadImage(img);
            l
        };

        let logo = unsafe { Texture2D::from_raw(logo_raw) };
        let logo_size = Vec2 { x: logo.width() as f32, y: logo.height() as f32 };

        let rand_x = rand::gen_range(0.0, window_size.x - logo_size.x);
        let rand_y = rand::gen_range(0.0, window_size.y - logo_size.y);

        Self {
            pos: Vec2{x: rand_x, y: rand_y},
            velocity: vec2(180.0, 180.0),
            window_size,
            logo_size,
            logo,
        }
    }

    fn update(&mut self) -> SelectedGame {
        let next = if let Some(KeyCode::Escape) = get_last_key_pressed() {
            SelectedGame::None
        } else {
            SelectedGame::Dvd
        };

        let dt = get_frame_time();

        self.window_size.x = screen_width();
        self.window_size.y = screen_height();

        self.pos = self
            .pos
            .clamp(vec2(0., 0.), self.window_size - self.logo_size)
            + self.velocity * dt;

        if self.pos.x + self.logo_size.x >= self.window_size.x || self.pos.x <= 0.0 {
            self.velocity.x *= -1.0;
        }
        if self.pos.y + self.logo_size.y >= self.window_size.y || self.pos.y <= 0.0 {
            self.velocity.y *= -1.0;
        }

        clear_background(WHITE);
        draw_texture(
            &self.logo,
            self.pos.x,
            self.pos.y,
            WHITE
        );

        next
    }
}
